app/controllers/generic_object_definition_controller.rb

```diff
@@ -151,13 +196,25 @@ def custom_button_group_node_info
     @custom_button_group_node = true
     @record = CustomButtonSet.find(from_cid(node.split("-").last))
     @right_cell_text = _("Custom Button Set %{record_name}") % {:record_name => @record.name}
+  rescue StandardError => _err
```

Comment:
This looks wrong, calling `custom_button_group_node_info` and `custom_button_node_info` and randomly switching to `root_node_info` on any exception sounds like a hack around something missing in `custom_button_group_node?` and `custom_button_node?`.

Reply:
How about we replace the word "hack" with "graceful node adjustment post delete"?
-- because that is what it is.

Try deleting a Custom Button or Custom Button Group node without the `rescue` and the exception that you will see is this -

```
F, [2017-11-06T10:39:37.535199 #9025] FATAL -- : Error caught: [ActiveRecord::RecordNotFound] Couldn't find CustomButton with 'id'=10000000000105
/Users/aparnakarve/.rvm/gems/ruby-2.4.1/gems/activerecord-5.0.6/lib/active_record/core.rb:173:in `find'
~/manageiq/plugins/manageiq-ui-classic/app/controllers/generic_object_definition_controller.rb:192:in `custom_button_node_info'
~/manageiq/plugins/manageiq-ui-classic/app/controllers/generic_object_definition_controller.rb:161:in `node_info'
~/manageiq/plugins/manageiq-ui-classic/app/controllers/generic_object_definition_controller.rb:23:in `show_list'
```

I can replace `StandardError` with `ActiveRecord::RecordNotFound`.
Although, I personally see no issues with `StandardError` error either because it gracefully transitions to the default node, which is the root node, in case of an error.

Reply:
Awaiting your reply here if we need to replace `StandardError` with `ActiveRecord::RecordNotFound` or leave the current `rescue` implementation as-is.

Reply:
>  but it looks like you know exactly when that error would happen (when deleting stuff)

No, that's the thing - the controller (`generic_object_definition_controller`) has no way to know when the error would happen, because the delete is being done by the Angular component - `genericObjectDefinitionToolbar`
The controller (`generic_object_definition_controller`), is merely in a "reaction" mode and simply reacts to the new status of the node (the non-existent status)

Reply:
Right.. so when does the error actually happen?

Is this after having deleted something, when you click on it in the UI? Or.. when?

I'm thinking you may need to notify the controller when deleting if it keeps some internal state whcih expects the item to exist.

Or at least to check if such a deletion hadn't ocurred before calling the wrong method?

Reply:
The error shows up when the angular component deletes the record and redirects back to the `show_list` url.
So, changing the `redirect-url` to include the query parameter `?id=root` should work to set the node to `root`

```ruby
%generic-object-definition-toolbar{'record-id'    => @record.id,
                                   'redirect-url' => "/generic_object_definition/show_list?id=root"}
```

Reply:
Ideally the above approach should work.
However, `node_info` has been programmed to honor only the session-based `x_node`, and therefore it would ignore `params[:id]`

What that means is, extra handling in  `show_list` to look for the presence of `params[:id] = root` and then adjust `x_node` based on that.

Reply:
Anyway, lets revisit this after I have done some of the other controller changes recommended below.
The controller landscape might change a little bit after I remove `GenericListMixin` as suggested here - https://github.com/ManageIQ/manageiq-ui-classic/pull/2569#issuecomment-342723073

Reply:
I had to add the `index` method because I removed `GenericListMixin`.

And that proved to be quite beneficial for this case since I could leverage the new `index` method for `redirect-url`.

The `rescue` is now gone.
Check out 59fcd90.